File: app/src/main/java/com/readdle/swiftjava/sample/SampleReference.java
package com.readdle.swiftjava.sample;

import com.readdle.codegen.anotation.SwiftCallbackFunc;
import com.readdle.codegen.anotation.SwiftDelegate;
import com.readdle.codegen.anotation.SwiftError;
import com.readdle.codegen.anotation.SwiftReference;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

@SwiftReference(importPackages = {"SampleAppCore"})
public class SampleReference {

    @SwiftReference(importPackages = {"SampleAppCore"})
    public static class SampleReferenceEnclose {


        // Swift JNI private native pointer
        private long nativePointer = 0L;

        // Swift JNI private constructor
        // Should be private. Don't call this constructor from Java!
        private SampleReferenceEnclose() {
        }

        // Swift JNI release method
        public native void release();

        @NonNull
        public static native SampleReferenceEnclose init();
    }

    @SwiftDelegate(importPackages = {"SampleAppCore"}, protocols = {"SampleBlockDelegate"})
    public interface SampleInterfaceDelegateAndroid {

        @SwiftCallbackFunc
        void onCall();

    }

    // Swift JNI private native pointer
    private long nativePointer = 0L;

    // Swift JNI private constructor
    // Should be private. Don't call this constructor from Java!
    private SampleReference() {
    }

    // Swift JNI release method
    public native void release();

    @NonNull
    public native SampleValue getRandomValue();

    public native void saveValue(@NonNull SampleValue value);

    public native void funcThrows() throws SwiftError;

    @Nullable
    public native SampleReference funcWithNil();

    @NonNull
    public static native SampleReference init();

    // TODO: Impossible to generate for now. Add extra check for JavaSwift protocol before casting to .javaObject()
    //@Nullable @SwiftFunc
    //public native SampleDelegateAndroid getDelegate();

    public native void setDelegate(SampleDelegateAndroid delegate);

    @NonNull
    public native Long tick();

    public native void tickWithBlock(@NonNull SampleInterfaceDelegateAndroid delegate);

    @NonNull
    public native Float floatCheck(@NonNull Float var1);

    @NonNull
    public native Double doubleCheck(@NonNull Double var1);

    @NonNull
    public native Exception exceptionCheck(@NonNull Exception var1);

    @NonNull
    public native SampleReferenceEnclose enclose(@NonNull SampleReferenceEnclose var1);

}

File: app/src/main/swift/Sources/SampleAppCore/SampleReference.swift
import Foundation

public protocol SampleDelegate: class {

    func setSampleValue(value: SampleValue?)

    func getSampleValue() -> SampleValue

    func funcWithNil() -> SampleValue?

    static func getTimestamp() -> Int64

    static func setTimestamp(_ value: Int64)
}

public protocol SampleBlockDelegate {

    func onCall()

}

private extension SampleDelegate {

    func timeStampTest() -> Int64 {
        let timestamp = Self.getTimestamp()
        Self.setTimestamp(timestamp)
        return timestamp
    }
}

public class SampleReferenceEnclose {

    public init() {

    }
}

public class SampleReference {

    // TODO: need some extra work for returning abstract type
    public weak var delegate: SampleDelegate?
    public var string: String = "str1"
    public static var staticString: String = "str1"

	public init() {
		
	}

    public func getRandomValue() -> SampleValue {
    	return SampleValue(str1: "1", str2: "2", str3: "3")
    }

    public func saveValue(_ value: SampleValue)  {
    	NSLog("SampleValue: \(value.str1), \(value.str2), \(value.str3)")
    }

    public func funcThrows() throws {
    	throw NSError(domain: "Error", code: 1)
    }

    public func tick() -> Int64 {
        if let delegate = self.delegate {
            let value = getRandomValue()
            delegate.setSampleValue(value: value)
            return delegate.timeStampTest()
        }
        return  -1
    }

    public func tickWithBlock(_ block: SampleBlockDelegate) {
        block.onCall()
    }

    public func funcWithNil() -> SampleValue? {
        delegate?.setSampleValue(value: nil)
        return delegate?.funcWithNil()
    }

    public func floatCheck(_ float: Float) -> Float {
        return float + 2.0
    }

    public func doubleCheck(_ double: Double) -> Double {
        return double + 2.0
    }

    public func exceptionCheck(_ var1: Error) -> Error {
        let domain = (var1 as! NSError).domain
        let code = (var1 as! NSError).code
        NSLog("Error: \(domain) \(code)")

        return var1
    }

    public func enclose(_ var1: SampleReferenceEnclose) -> SampleReferenceEnclose {
        return var1
    }


    public func funcWithBlock(completion: ((Error?, String?) -> String?)?) -> String? {
        return completion?(nil, "123") ?? "null"
    }

}
